torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+//      after the first call to a function.
+//
+//  4) Storing time relative to the start of profiling.
+//      In general profiling is short lived. Storing an entire int64_t just to
+//      record that a handful of microseconds have passed is not a good use of
+//      bits. So instead, we record the time since profiling began. We can
+//      fit over an hour into a uint32_t which is far longer than the profiler
+//      should ever run for a continuous period.
+//
+// With these tricks we can pack all of the above into a single 8 byte word.
+// The second word is case dependent.
+//
+// One obvious question is: why manually tag the union rather than using a
+// `std::variant`? (Or `c10::variant`, as it were.) The answer is that due
+// to alignment the tag would have to be packed with the union data and
+// `RawEvent` would grow to three words. (Not just 50% bigger, but also less
+// cache friendly.)
+struct RawEvent {
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx)
+            : tag_(static_cast<uint8_t>(tag)),
+              thread_id_(ctx->thread_id_),
+              lasti_(static_cast<uint16_t>(lasti)),
+              misc_() {
+        int64_t t = now() - ctx->initial_us_;
+        t_ = static_cast<uint32_t>(t);
+
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(lasti <= std::numeric_limits<uint16_t>::max());
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t <= std::numeric_limits<uint32_t>::max());
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyCodeObject* f_code)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kPy_Call);
+        misc_.f_code_ = f_code;
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyObject* arg)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kC_Call);
+        misc_.arg_ = arg;
+    }
+
+    uint8_t tag_;
+    uint8_t thread_id_;
+    uint16_t lasti_;
+    uint32_t t_;
+    union {
+        // TraceTag::kPy_Call
+        PyCodeObject* f_code_;
+
+        // TraceTag::kC_Call
+        PyObject* arg_;
+
+        // TraceTag::kPy_Return
+        // TraceTag::kC_Return
+        // ** Unused (placeholder) **
+        void* null_;
+    } misc_;
+
+    TraceTag tag() const {
+        return static_cast<TraceTag>(tag_);
+    }
+
+    int lasti() const {
+        // f_lasti is positive, with one exception: CPython intializes frames
+        // with `f_lasti = -1`. We don't want to give up half of the range by
+        // switching to int16_t. So instead we do the fast (underflowing) cast
+        // in the ctor, and rectify the value in this accessor which should
+        // only be called during trace post processing.
+        return lasti_ == std::numeric_limits<uint16_t>::max()
+            ? (int)(-1)
+            : (int)lasti_;
+    }
+};
+
+static_assert(sizeof(RawEvent) == 16, "RawEvent should be exactly 16 bytes.");
```

Comment:
the assertion already screams that "RawEvent should be exactly 16 bytes." 😃 probably add high level reasoning why this should stay e.g. to fit within cache line

Reply:
Done.